rust/headless-client/src/ipc_service.rs

```diff
@@ -438,36 +436,47 @@ impl<'a> Handler<'a> {
                     .context("Failed to send `ConnectResult`")?
             }
             ClientMsg::Disconnect => {
-                if let Some(connlib) = self.connlib.take() {
-                    connlib.disconnect();
-                    self.dns_controller.deactivate()?;
-                } else {
+                let Some(session) = self.session.take() else {
                     tracing::error!("Error - Got Disconnect when we're already not connected");
-                }
+                    return Ok(());
+                };
+
+                // Identical to dropping it, but looks nicer.
+                session.connlib.disconnect();
+                self.dns_controller.deactivate()?;
             }
             ClientMsg::ReloadLogFilter => {
                 let filter = spawn_blocking(get_log_filter).await??;
                 self.log_filter_reloader.reload(filter)?;
             }
             ClientMsg::Reset => {
-                if self.tunnel_is_ready() {
-                    self.connlib.as_mut().context("No connlib session")?.reset();
-                } else {
-                    tracing::debug!("Ignoring redundant reset");
+                if self.last_connlib_start_instant.is_some() {
+                    tracing::debug!("Ignoring reset since we're still signing in");
+                    return Ok(());
                 }
+                let Some(session) = self.session.as_ref() else {
+                    tracing::debug!("Cannot reset if we're signed out");
+                    return Ok(());
+                };
+
+                session.connlib.reset();
             }
             ClientMsg::SetDns(resolvers) => {
+                let Some(session) = self.session.as_ref() else {
+                    tracing::debug!("Cannot set DNS resolvers if we're signed out");
+                    return Ok(());
+                };
+
                 tracing::debug!(?resolvers);
-                self.connlib
-                    .as_mut()
-                    .context("No connlib session")?
-                    .set_dns(resolvers)
+                session.connlib.set_dns(resolvers);
             }
             ClientMsg::SetDisabledResources(disabled_resources) => {
-                self.connlib
-                    .as_mut()
-                    .context("No connlib session")?
-                    .set_disabled_resources(disabled_resources);
+                let Some(session) = self.session.as_ref() else {
+                    tracing::warn!("Cannot set disabled resources if we're signed out");
```

Comment:
Have they been saved to the settings at this point? If yes, I don't think this warrants a `warn`.